File: Editorial/36_Valid_Sudoku/TypeScript/2_Array_of_Fixed_Length.ts
function isValidSudoku(board: string[][]): boolean {
    let N = 9;
    let rows: number[][] = Array.from({ length: N }, () => Array(N).fill(0));
    let cols: number[][] = Array.from({ length: N }, () => Array(N).fill(0));
    let boxes: number[][] = Array.from({ length: N }, () => Array(N).fill(0));
    for (let r = 0; r < N; r++) {
        for (let c = 0; c < N; c++) {
            if (board[r][c] == ".") {
                continue;
            }
            let pos: number = parseInt(board[r][c]) - 1;
            if (rows[r][pos] == 1) {
                return false;
            }
            rows[r][pos] = 1;
            if (cols[c][pos] == 1) {
                return false;
            }
            cols[c][pos] = 1;
            let idx: number = Math.floor(r / 3) * 3 + Math.floor(c / 3);
            if (boxes[idx][pos] == 1) {
                return false;
            }
            boxes[idx][pos] = 1;
        }
    }
    return true;
}

File: Editorial/36_Valid_Sudoku/C/2_Array_of_Fixed_Length.c
bool isValidSudoku(char** board, int boardSize, int* boardColSize) {
    int N = 9;
    int rows[N][N];
    memset(rows, 0, sizeof(rows));
    int cols[N][N];
    memset(cols, 0, sizeof(cols));
    int boxes[N][N];
    memset(boxes, 0, sizeof(boxes));
    for (int r = 0; r < N; r++) {
        for (int c = 0; c < N; c++) {
            if (board[r][c] == '.') {
                continue;
            }
            int pos = board[r][c] - '1';
            if (rows[r][pos] == 1) {
                return false;
            }
            rows[r][pos] = 1;
            if (cols[c][pos] == 1) {
                return false;
            }
            cols[c][pos] = 1;
            int idx = (r / 3) * 3 + c / 3;
            if (boxes[idx][pos] == 1) {
                return false;
            }
            boxes[idx][pos] = 1;
        }
    }
    return true;
}
